---
title: "02_Pseudobulk_DEseq2"
output: html_notebook
---

```{r setup, include = FALSE, echo = FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(scater)
library(Seurat)
library(tidyverse)
library(cowplot)
library(Matrix.utils)
library(dplyr)
library(magrittr)
library(Matrix)
library(purrr)
library(reshape2)
library(S4Vectors)
library(tibble)
library(SingleCellExperiment)
library(pheatmap)
library(apeglm)
library(png)
library(DESeq2)
library(RColorBrewer)
library(biomaRt)
library(clusterProfiler)
library(org.Mm.eg.db)
library(DT)
library(ggvenn)

knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)

```

```{r prepare data}
# Bring in Seurat object
seurat <- readRDS("~/Desktop/Projects/La_Spada/SCA7/snRNA-seq/Seurat/Objects/SCA7.8wk.cleaned.rds")
seurat

# First function extracts counts from seurat object
# So we want to prepare a list of seurat objects, subset by each cell type
cell.type.table <- table(Idents(seurat),seurat$Type) %>% as.data.frame.matrix() %>% mutate(total = SCA7 + WT) %>% arrange(-total)
cell.type.table # shows each cell type and numbers for SCA7 & WT
celltypes <- row.names(cell.type.table)

# loop over each celltype and store in a seurat.list
seurat.list <- list() # create empty list to hold outputs

for (i in 1:13) {
  celltype <- row.names(cell.type.table)[i] # pull celltype from list
  subset <- subset(seurat, idents = celltype) # subset new seurat object
  
  seurat.list[[i]] <- subset
}

seurat.list[[1]]
```

```{r create function for extracting counts and metadata}

extractCounts <- function(seurat){
  # Extract raw counts and metadata to create SingleCellExperiment object
  counts <- seurat@assays$RNA@counts 
  metadata <- seurat@meta.data
  
  # Set up metadata as desired for aggregation and DE analysis
  metadata$cluster_id <- factor(seurat@active.ident) # contains Named Cell Types which can be used to split DE by celltype
  metadata$Type <- factor(metadata$Type)
  metadata$MULTIseq_group <- factor(metadata$MULTIseq_group)
  metadata$Sex <- factor(metadata$Sex)
  
  # Create single cell experiment object
  sce <- SingleCellExperiment(assays = list(counts = counts), 
                             colData = metadata)
  
  return(sce)
}

sce.list <- lapply(seurat.list, extractCounts)
sce.list
```

## Sample level metadata

```{r}
# Write function to convert sce object to DEseq object
convertDDS <- function(sce){
  ## Create a numeric vector of number of cells per sample
  n_cells <- as.numeric(table(sce$MULTIseq_group))
  
  # Create a named vector of sample names
  sids <- purrr::set_names(levels(factor(sce$MULTIseq_group)))
  ## Determine how to reorder the samples (rows) of the metadata to match the order of sample names in sids vector
  m <- match(sids, sce$MULTIseq_group)
  
  
  ## Create the sample level metadata by combining the reordered metadata with the number of cells corresponding to each sample.
  metadata <- data.frame(colData(sce)[m, ], 
                   n_cells,
                   row.names = NULL)
  metadata <- metadata %>% dplyr::select(-"cluster_id", -"orig.ident", -"nCount_RNA", -"nFeature_RNA", -"percent.mt", -"RNA_snn_res.0.8", -"seurat_clusters")
  
  rownames(metadata) <- metadata$MULTIseq_group
  
  ## Count aggregation to sample level
  
  # Aggregate the counts per sample_id and cluster_id
  # Subset metadata to only include the cluster and sample IDs to aggregate across
  groups <- colData(sce)[, c("MULTIseq_group")]
  
  # Aggregate across cluster-sample groups
  pb <- aggregate.Matrix(t(counts(sce)), 
                         groupings = groups, 
                         fun = "sum") 
  
  # DESeq2 analysis
  
  counts <- as.data.frame(t(as.matrix(pb)))
  
  # Define comparison group
  group <- factor(metadata$Type)
  
  # Create DESeq2 object
  dds <- DESeqDataSetFromMatrix(counts, 
                                colData = metadata, 
                                design = ~Sex+Type)
  
  dds$Type <- relevel(dds$Type, ref = "WT")
  return(dds)
}

dds.list <- lapply(sce.list, convertDDS)
dds.list
```

## Sample level metadata with Sex Interaction variable

```{r}
# Write function to convert sce object to DEseq object
convertDDS.sex.interaction <- function(sce){
  ## Create a numeric vector of number of cells per sample
  n_cells <- as.numeric(table(sce$MULTIseq_group))
  
  # Create a named vector of sample names
  sids <- purrr::set_names(levels(factor(sce$MULTIseq_group)))
  ## Determine how to reorder the samples (rows) of the metadata to match the order of sample names in sids vector
  m <- match(sids, sce$MULTIseq_group)
  
  
  ## Create the sample level metadata by combining the reordered metadata with the number of cells corresponding to each sample.
  metadata <- data.frame(colData(sce)[m, ], 
                   n_cells,
                   row.names = NULL)
  metadata <- metadata %>% dplyr::select(-"cluster_id", -"orig.ident", -"nCount_RNA", -"nFeature_RNA", -"percent.mt", -"RNA_snn_res.0.8", -"seurat_clusters")
  
  rownames(metadata) <- metadata$MULTIseq_group
  
  ## Count aggregation to sample level
  
  # Aggregate the counts per sample_id and cluster_id
  # Subset metadata to only include the cluster and sample IDs to aggregate across
  groups <- colData(sce)[, c("MULTIseq_group")]
  
  # Aggregate across cluster-sample groups
  pb <- aggregate.Matrix(t(counts(sce)), 
                         groupings = groups, 
                         fun = "sum") 
  
  # DESeq2 analysis
  
  counts <- as.data.frame(t(as.matrix(pb)))
  
  # Define comparison group
  group <- factor(metadata$Type)
  
  # Create DESeq2 object
  dds <- DESeqDataSetFromMatrix(counts, 
                                colData = metadata, 
                                design = ~Sex+Type+Sex:Type)
  
  dds$Type <- relevel(dds$Type, ref = "WT")
  return(dds)
}

dds.sex.list <- lapply(sce.list, convertDDS.sex.interaction)
dds.sex.list
```

```{r DE 1}
# Create function to run DEseq and return full results as tibble
runDEseq <- function(dds){
  # Run DESeq2 differential expression analysis
  dds1 <- estimateSizeFactors(dds)
  dds1 <- estimateDispersions(dds1)
  dds1 <- nbinomWaldTest(dds1)
  
  # Output results of Wald test for contrast for SCA7 vs WT
  res <- results(dds1, 
                 contrast = c("Type", "SCA7", "WT"),
                 alpha = 0.05)
  
  res <- lfcShrink(dds1, 
                   coef = "Type_SCA7_vs_WT",
                   res = res)
  
  # Turn the results object into a tibble for use with tidyverse functions
  res_tbl <- res %>%
          data.frame() %>%
          rownames_to_column(var="gene") %>%
          as_tibble()
  
  res_tbl <- res_tbl %>% arrange(-log2FoldChange)
}

res.list <- lapply(dds.list, runDEseq)
res.list[[1]] # full results list is hard to handle, filter for significant & logFC > 0.5

# add column denoting celltype to full results
for(i in 1:13){
  res.list[[i]] <- res.list[[i]] %>% mutate(celltype = celltypes[i])
}

res.list

# save full results
res.list %>% saveRDS(file = "~/DEG-output-path/8wk-pseudobulk-replicate-full.rds")
```

```{r}
# Filter results for significant & logFC > 0.5
filterSig <- function(res){
  filter(res, abs(log2FoldChange) > 0.5 & padj < 0.05)
}

res.sig <- lapply(res.list, filterSig)
res.sig

cell.type.table

# generate vector of celltypes
celltypes <- row.names(cell.type.table)
res.sig[[i]] %>% mutate(celltype = celltypes[1])

# add column denoting celltype to sig results
for(i in 1:13){
  res.sig[[i]] <- res.sig[[i]] %>% mutate(celltype = celltypes[i])
}

res.sig
```

```{r}
# view names of interaction terms
dd.test <- dds.list[[1]]

dds1 <- estimateSizeFactors(dd.test)
dds1 <- estimateDispersions(dds1)
dds1 <- nbinomWaldTest(dds1)

resultsNames(dds1)

resultsInteraction <- results(dds1, name="SexMale.TypeSCA7")
resultsInteraction <- lfcShrink(dds1, 
                   coef = "SexMale.TypeSCA7",
                   res = resultsInteraction)

resultsInteraction %>% data.frame() %>% rownames_to_column(var="gene") %>%
          as_tibble() %>% arrange(pvalue) ## no significant genes varying by SCA7:Sex in Purkinje cells

# Create function to run DEseq and return sex interaction results as tibble
runDEseq.sex <- function(dds){
  # Run DESeq2 differential expression analysis
  dds1 <- estimateSizeFactors(dds)
  dds1 <- estimateDispersions(dds1)
  dds1 <- nbinomWaldTest(dds1)
  
  # Output results of Wald test for contrast for SCA7 vs WT
  res.sex <- results(dds1, 
                 name="SexMale.TypeSCA7")
  
  res.sex <- lfcShrink(dds1, 
                   coef = "SexMale.TypeSCA7",
                   res = res.sex)
  
  # Turn the results object into a tibble for use with tidyverse functions
  res_tbl_sex <- res.sex %>%
          data.frame() %>%
          rownames_to_column(var="gene") %>%
          as_tibble()
  
  res_tbl_sex <- res_tbl_sex %>% arrange(padj)
}

res.list.sex <- lapply(dds.sex.list, runDEseq.sex)
res.list.sex # zero genes with significant sex interaction in any cell type
```


```{r}
# only 6 cell types have significant DE based on cutoffs
# bind all 6 into single long df
Celltype.deg.long <- rbind(res.sig[[1]],res.sig[[2]],res.sig[[4]],res.sig[[7]],res.sig[[9]],res.sig[[10]])
Celltype.deg.long

# Extract CellTypes and order by # of DEGs
Celltypes.order <- Celltype.deg.long$celltype %>% table() %>% as.data.frame() %>% arrange(-Freq)
Celltypes.order <- Celltypes.order$.
Celltypes.order

# need to make celltype an ordered factor so it does not get alphabetical sorted
Celltype.deg.long$celltype <- factor(Celltype.deg.long$celltype, levels = Celltypes.order)
Celltype.deg.long

# plot DEGs
p1 <- ggplot(arrange(Celltype.deg.long,-padj), aes(x=celltype, y=log2FoldChange, color=-log(padj))) + 
  geom_jitter(height = 0, width = 0.15) +
  scale_color_gradient(low = "blue", high = "red") +
  xlab('') + ylab('log2FC') +
  #ylim(-3,5) +
  labs(title = "Pseudobulk DESeq2 DEGs per Cell Type") +
  theme_bw() +
  theme(
    legend.position = 'right',
    legend.background = element_rect(),
    plot.title = element_text(angle = 0, size = 14, face = 'bold', vjust = 1, hjust = 0.5),

    axis.text.x = element_text(angle = 90, size = 10, face = 'bold', vjust = 0.5),
    axis.text.y = element_text(angle = 0, size = 12, face = 'bold', vjust = 0.5),
    axis.title = element_text(size = 12, face = 'bold'),
    axis.title.x = element_text(size = 12, face = 'bold'),
    axis.title.y = element_text(size = 12, face = 'bold'),
    axis.line = element_line(colour = 'black'),

    #Legend
    legend.key = element_blank(), # removes the border
    legend.key.size = unit(0.75, "cm"), # Sets overall area/size of the legend
    legend.text = element_text(size = 12, face = "bold"), # Text size
    title = element_text(size = 12, face = "bold")) 
p1
```

Repeat for 5wk Data

```{r}
# Bring in 5wk Seurat object
seurat <- readRDS("~/Desktop/Projects/La_Spada/SCA7/snRNA-seq/Seurat/Objects/SCA7.5wk.cleaned.rds")
seurat

# First function extracts counts from seurat object
# So we want to prepare a list of seurat objects, subset by each cell type
cell.type.table <- table(Idents(seurat),seurat$Type) %>% as.data.frame.matrix() %>% mutate(total = SCA7 + WT) %>% arrange(-total)
cell.type.table # shows each cell type and numbers for SCA7 & WT

# loop over each celltype and store in a seurat.list
seurat.list <- list() # create empty list to hold outputs

for (i in 1:13) {
  celltype <- row.names(cell.type.table)[i] # pull celltype from list
  subset <- subset(seurat, idents = celltype) # subset new seurat object
  
  seurat.list[[i]] <- subset
}

seurat.list[[1]]
```

```{r}
extractCounts <- function(seurat){
  # Extract raw counts and metadata to create SingleCellExperiment object
  counts <- seurat@assays$RNA@counts 
  metadata <- seurat@meta.data
  
  # Set up metadata as desired for aggregation and DE analysis
  metadata$cluster_id <- factor(seurat@active.ident) # contains Named Cell Types which can be used to split DE by celltype
  metadata$Type <- factor(metadata$Type)
  metadata$MULTIseq_group <- factor(metadata$MULTIseq_group)
  metadata$Sex <- factor(metadata$Sex)
  
  # Create single cell experiment object
  sce <- SingleCellExperiment(assays = list(counts = counts), 
                             colData = metadata)
  
  return(sce)
}

sce.list <- lapply(seurat.list, extractCounts)
sce.list
```

## Sample level metadata

```{r}
# Write function to convert sce object to DEseq object
convertDDS <- function(sce){
  ## Create a numeric vector of number of cells per sample
  n_cells <- as.numeric(table(sce$MULTIseq_group))
  
  # Create a named vector of sample names
  sids <- purrr::set_names(levels(factor(sce$MULTIseq_group)))
  ## Determine how to reorder the samples (rows) of the metadata to match the order of sample names in sids vector
  m <- match(sids, sce$MULTIseq_group)
  
  
  ## Create the sample level metadata by combining the reordered metadata with the number of cells corresponding to each sample.
  metadata <- data.frame(colData(sce)[m, ], 
                   n_cells,
                   row.names = NULL)
  metadata <- metadata %>% dplyr::select(-"cluster_id", -"orig.ident", -"nCount_RNA", -"nFeature_RNA", -"percent.mt", -"RNA_snn_res.0.8", -"seurat_clusters")
  
  rownames(metadata) <- metadata$MULTIseq_group
  
  ## Count aggregation to sample level
  
  # Aggregate the counts per sample_id and cluster_id
  # Subset metadata to only include the cluster and sample IDs to aggregate across
  groups <- colData(sce)[, c("MULTIseq_group")]
  
  # Aggregate across cluster-sample groups
  pb <- aggregate.Matrix(t(counts(sce)), 
                         groupings = groups, 
                         fun = "sum") 
  
  # DESeq2 analysis
  
  counts <- as.data.frame(t(as.matrix(pb)))
  
  # Define comparison group
  group <- factor(metadata$Type)
  
  # Create DESeq2 object
  dds <- DESeqDataSetFromMatrix(counts, 
                                colData = metadata, 
                                design = ~Sex+Type)
  
  dds$Type <- relevel(dds$Type, ref = "WT")
  return(dds)
}

dds.list <- lapply(sce.list, convertDDS)
dds.list
```

```{r}
# Create function to run DEseq and return full results as tibble
runDEseq <- function(dds){
  # Run DESeq2 differential expression analysis
  dds1 <- estimateSizeFactors(dds)
  dds1 <- estimateDispersions(dds1)
  dds1 <- nbinomWaldTest(dds1)
  
  # Output results of Wald test for contrast for SCA7 vs WT
  res <- results(dds1, 
                 contrast = c("Type", "SCA7", "WT"),
                 alpha = 0.05)
  
  res <- lfcShrink(dds1, 
                   coef = "Type_SCA7_vs_WT",
                   res = res)
  
  # Turn the results object into a tibble for use with tidyverse functions
  res_tbl <- res %>%
          data.frame() %>%
          rownames_to_column(var="gene") %>%
          as_tibble()
  
  res_tbl <- res_tbl %>% arrange(-log2FoldChange)
}

res.list <- lapply(dds.list, runDEseq)

# generate vector of celltypes
celltypes <- row.names(cell.type.table)

# add column denoting celltype to all results
for(i in 1:13){
  res.list[[i]] <- res.list[[i]] %>% mutate(celltype = celltypes[i])
}

res.list[[1]] # full results list is hard to handle, filter for significant & logFC > 0.5
res.list
```

```{r}
# Filter results for significant & logFC > 0.5
filterSig <- function(res){
  filter(res, abs(log2FoldChange) > 0.5 & padj < 0.05)
}

res.sig <- lapply(res.list, filterSig)
res.sig
```

```{r}
# only Purkinje Cells have significant DE based on cutoffs
Celltype.5wk.deg.long <- rbind(res.sig[[1]])
Celltype.5wk.deg.long

# Extract CellTypes and order by # of DEGs
Celltypes.order <- Celltype.5wk.deg.long$celltype %>% table() %>% as.data.frame() %>% arrange(-Freq)
Celltypes.order <- Celltypes.order$.
Celltypes.order

# need to make celltype an ordered factor so it does not get alphabetical sorted
Celltype.5wk.deg.long$celltype <- factor(Celltype.5wk.deg.long$celltype, levels = Celltypes.order)
Celltype.5wk.deg.long

# Plot DEGs
p2 <- ggplot(arrange(Celltype.5wk.deg.long,-padj), aes(x=celltype, y=log2FoldChange, color=-log(padj))) + 
  geom_jitter(height = 0, width = 0.15) +
  scale_color_gradient(low = "blue", high = "red") +
  xlab('') + ylab('log2FC') +
  #ylim(-3,5) +
  labs(title = "Pseudobulk DESeq2 DEGs per Cell Type") +
  theme_bw() +
  theme(
    legend.position = 'right',
    legend.background = element_rect(),
    plot.title = element_text(angle = 0, size = 14, face = 'bold', vjust = 1, hjust = 0.5),

    axis.text.x = element_text(angle = 90, size = 10, face = 'bold', vjust = 0.5),
    axis.text.y = element_text(angle = 0, size = 12, face = 'bold', vjust = 0.5),
    axis.title = element_text(size = 12, face = 'bold'),
    axis.title.x = element_text(size = 12, face = 'bold'),
    axis.title.y = element_text(size = 12, face = 'bold'),
    axis.line = element_line(colour = 'black'),

    #Legend
    legend.key = element_blank(), # removes the border
    legend.key.size = unit(0.75, "cm"), # Sets overall area/size of the legend
    legend.text = element_text(size = 12, face = "bold"), # Text size
    title = element_text(size = 12, face = "bold")) 

p2
```
